### ISSUE TITLE
Retries failed with "tls: use of closed connection" error

### ISSUE THREAD

**Issue by bluekeyes** [git-lfs/git-lfs#3929]
We're occasionally seeing failures downloading LFS files from GitHub Enterprise (2.18.5 and 2.19.2) where a failed request is retried, but then the retry fails due to a `tls: use of closed connection` error. Here's a `GIT_TRACE=1` log from one occurrence:

```
04:34:43.194170 trace git-lfs: exec: git '-c' 'filter.lfs.smudge=cat' '-c' 'filter.lfs.clean=cat' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'rev-parse' 'HEAD' '--symbolic-full-name' 'HEAD'
Downloading path/to/file.png (136 KB)
04:34:43.200267 trace git-lfs: tq: running as batched queue, batch size of 100
04:34:43.200331 trace git-lfs: tq: sending batch of size 1
04:34:43.200446 trace git-lfs: run_command: ssh -- git@github.company.domain git-lfs-authenticate org/repository.git download
04:35:01.175099 trace git-lfs: api: batch 1 files
04:35:01.175492 trace git-lfs: HTTP: POST https://github.company.domain/_lfs/org/repository/objects/batch
04:35:02.002612 trace git-lfs: HTTP: 200
04:35:02.002782 trace git-lfs: HTTP: {"objects":[{"oid":"fdb49757db5ee89cd90af00c1671c23912e11c60237d743a6c1fa1487006edfa","size":136253,"actions":{"download":{"href":"https://media.github.company.domain/lfs/58863/objects/fdb49757db5ee89cd90af00c1671c23912e11c60237d743a6c1fa1487006edfa","header":{"Authorization":"RemoteAuth REDACTED"}}}}]}
04:35:02.002881 trace git-lfs: tq: starting transfer adapter "basic"
04:35:02.003394 trace git-lfs: HTTP: GET https://media.github.company.domain/lfs/58863/objects/fdb49757db5ee89cd90af00c1671c23912e11c60237d743a6c1fa1487006edfa
04:35:17.093724 trace git-lfs: tq: retrying object fdb49757db5ee89cd90af00c1671c23912e11c60237d743a6c1fa1487006edfa: LFS: Get https://media.github.company.domain/lfs/58863/objects/fdb49757db5ee89cd90af00c1671c23912e11c60237d743a6c1fa1487006edfa: EOF
04:35:17.093762 trace git-lfs: tq: enqueue retry #1 for "fdb49757db5ee89cd90af00c1671c23912e11c60237d743a6c1fa1487006edfa" (size: 136253)
04:35:17.093800 trace git-lfs: tq: sending batch of size 1
04:35:17.093937 trace git-lfs: ssh cache: git@github.company.domain git-lfs-authenticate org/repository.git download
04:35:17.093991 trace git-lfs: api: batch 1 files
04:35:17.094092 trace git-lfs: HTTP: POST https://github.company.domain/_lfs/org/repository/objects/batch
04:35:17.136403 trace git-lfs: api error: Post https://github.company.domain/_lfs/org/repository/objects/batch: tls: use of closed connection
Error downloading object: path/to/file.png (fdb4975): Smudge error: Error downloading path/to/file.png (fdb49757db5ee89cd90af00c1671c23912e11c60237d743a6c1fa1487006edfa): batch response: Post https://github.company.domain/_lfs/org/repository/objects/batch: tls: use of closed connection
```

When this happens, it seems that `git-lfs` stops retrying and fails, even though I believe the default number of retries is 8.

This may be a regression in 2.9.0. I believe we've only seen this error since upgrading our CI infrastructure, but I haven't confirmed that we were definitely running an older version before.

Here's the output of `git lfs env` (run on a CI machine, but outside of a repository):

```
git-lfs/2.9.0 (GitHub; linux amd64; go 1.13.1)
git version 2.7.4

LocalWorkingDir=
LocalGitDir=
LocalGitStorageDir=
LocalMediaDir=lfs/objects
LocalReferenceDirs=
TempDir=lfs/tmp
ConcurrentTransfers=3
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file
UploadTransfers=basic,lfs-standalone-file
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

Please let me know if there's any additional debug information that would be helpful. The problem reproduces multiple times a day, so we should be able to collect more examples.

**Comment by bk2204** [comment]
Hey,

Sorry to hear you're having trouble.  I suspect this may be related to things using HTTP/2 in 2.9.0, since we may not have reused connections in the same way before.  If the server you're using is fronted by nginx, then this is likely the reason.

Could you run the operation with `GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1` and pull out the request and response for the `/objects/batch` endpoint (removing any credentials from the `Authorization` headers)?  I'd like to see if we're using HTTP/1.1 or HTTP/2, since that will help me track down where we're getting this message from.

If it looks like it's HTTP/2. you can try with 2.9.1-pre (in #3928 ) or 2.9.1 (when it comes out Monday) and set `http.version` to `HTTP/1.1` and see if that helps as a workaround.

**Comment by bluekeyes** [comment]
Thanks @bk2204, I'll try to collect the additional debug information from the client today.

If it helps as a starting point, the HAProxy logs from GitHub suggest that we're using HTTP/1.1:

```
xx.x.xxx.xxx:44898 [22/Nov/2019:04:35:01.179] https_protocol~ lfs_server/localhost 0/0/1/168/169 200 1346 - - ---- 153/144/2/2/0 0/0 {github.company.domain||git-lfs/2.9.0 (GitHub; linux amd64; go 1.13.1)|TLSv1.2} "POST /_lfs/org/repository/objects/batch HTTP/1.1"
```

While our GHE instance is behind an AWS classic load balancer, it doesn't support HTTP/2 and we're also using the lower-level SSL listener instead of an HTTPS listener. I don't think there are any other proxies in the traffic path, but I'll wait for the client logs to confirm.

**Comment by bk2204** [comment]
Okay, that sounds like it's probably fine.  I have a patch I think may fix things; I'll open a draft PR with it and ping you on it.  You're welcome to download the artifacts from the CI build (or build your own) and test them on your systems to see if they fix the problem and let us know.

**Comment by bluekeyes** [comment]
Thanks, I've tried deploying one of the CI artifacts from #3930, so I'll update this if we still see the closed connection errors.

**Comment by bluekeyes** [comment]
Running with one of the CI artifacts appears to fix the problem, and reverting back to the 2.9.0 release causes failures to reappear.

**Comment by bk2204** [comment]
Great.  Once I get the CI fix merged, I'll move #3930 along as a full pull request.

**Comment by bk2204** [comment]
Hey,

#3930 has been merged and will be included in v2.9.2, which is due to be released Thursday, so I'm going to close this issue.
